pyomo/contrib/cp/repn/docplex_writer.py

```diff
@@ -0,0 +1,1107 @@
+                                                 arg in args[1:]), 1),
+                                       _get_int_valued_expr(args[0])))
+
+def _handle_at_least_node(visitor, node, *args):
+    return (_GENERAL, cp.greater_or_equal(cp.count((_get_bool_valued_expr(arg)
+                                                    for arg in args[1:]), 1),
+                                          _get_int_valued_expr(args[0])))
+
+## CallExpression handllers
+
+def _before_call_dispatcher(visitor, node, *args):
+    if len(args) == 2:
+        return _handle_inequality_node(visitor, node, args[0], args[1])
+    else: # a delay is also specified
+        lhs = _handle_sum_node(visitor, node, args[0], args[2])
+        return _handle_inequality_node(visitor, node, lhs, args[1])
+
+def _after_call_dispatcher(visitor, node, *args):
+    if len(args) == 2:
+        return _handle_inequality_node(visitor, node, args[1], args[0])
+    else: # delay is also specified
+        lhs = _handle_sum_node(visitor, node, args[1], args[2])
+        return _handle_inequality_node(visitor, node, lhs, args[0])
+
+def _at_call_dispatcher(visitor, node, *args):
+    if len(args) == 2:
+        return _handle_equality_node(visitor, node, args[0], args[1])
+    else: # a delay is also specified
+        rhs = _handle_sum_node(visitor, node, args[1], args[2])
+        return _handle_equality_node(visitor, node, args[0], rhs)
+
+_call_dispatchers = {
+    _DEFERRED_BEFORE: _before_call_dispatcher,
+    _DEFERRED_AFTER: _after_call_dispatcher,
+    _DEFERRED_AT: _at_call_dispatcher,
+    _IMPLIES: _handle_implication_node,
+    _LAND: _handle_and_node,
+    _LOR: _handle_or_node,
+    _XOR: _handle_xor_node,
+    _EQUIVALENT_TO: _handle_equivalence_node,
+}
+
+def _handle_call(visitor, node, *args):
+    return _call_dispatchers[args[0][0]](visitor, node, args[0][1], *args[1:])
+
+##
+# Scheduling
+##
+
+if docplex_available:
+    _before_dispatchers = {
+        (_START_TIME, _START_TIME): cp.start_before_start,
+        (_START_TIME, _END_TIME): cp.start_before_end,
+        (_END_TIME, _START_TIME): cp.end_before_start,
+        (_END_TIME, _END_TIME): cp.end_before_end,
+    }
+    _at_dispatchers = {
+        (_START_TIME, _START_TIME): cp.start_at_start,
+        (_START_TIME, _END_TIME): cp.start_at_end,
+        (_END_TIME, _START_TIME): cp.end_at_start,
+        (_END_TIME, _END_TIME): cp.end_at_end
+    }
+    _time_point_dispatchers = {
+        _START_TIME: cp.start_of,
+        _END_TIME: cp.end_of,
+        _GENERAL: lambda x: x,
+        _ELEMENT_CONSTRAINT: lambda x: x,
+    }
+
+def _handle_before_expression_node(visitor, node, time1, time2, delay):
+    t1 = (_GENERAL, _time_point_dispatchers[time1[0]](time1[1]))
+    t2 = (_GENERAL, _time_point_dispatchers[time2[0]](time2[1]))
+    lhs = _handle_sum_node(visitor, None, t1, delay)
+    if (time1[0] in {_GENERAL, _ELEMENT_CONSTRAINT} or
+        time2[0] in {_GENERAL, _ELEMENT_CONSTRAINT}):
```

Comment:
It is slightly more efficient to only create this set once and cache it at the module scope